orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc

```diff
@@ -0,0 +1,550 @@
+  // nodes in candidate_input are the input nodes to the subgraph constituted by candidate_edges
+  // input edges of nodes in candidate_input should be in candidate_edges or to be added Reshape + Gather
+  std::unordered_set<Node*> candidate_input;
+  // nodes in candidate_output are the output nodes from the subgraph constituted by candidate_edges
+  // input edges of nodes in candidate_output, if in candidate_edges, should be added GatherGrad + Reshape
+  std::unordered_set<Node*> candidate_output;
+  std::queue<Node*> to_visit;
+  std::unordered_set<Node*> visited;
+
+  // Find the valid embedding node
+  for (auto node_index : node_topology_list) {
+    auto& node = *graph.GetNode(node_index);
+    ORT_RETURN_IF_ERROR(Recurse(node, modified, graph_level, logger));
+
+    if (IsATenEmbedding(&node) &&
+        graph_utils::IsSupportedProvider(node, GetCompatibleExecutionProviders()) &&
+        node.InputDefs().size() >= 3 &&
+        node.InputDefs()[2]->Exists() &&
+        graph_utils::IsConstantInitializer(graph, node.InputDefs()[2]->Name()) &&
+        node.InputDefs()[1]->Exists() &&
+        graph_utils::IsGraphInput(graph, node.InputDefs()[1]) &&
+        node.InputDefs()[1]->Shape()->dim(0).has_dim_param() &&
+        node.InputDefs()[1]->Shape()->dim(1).has_dim_param()) {
+      const TensorProto* tensor_proto = graph_utils::GetConstantInitializer(graph, node.InputDefs()[2]->Name());
+      if (tensor_proto != nullptr &&
+          tensor_proto->dims_size() == 0 &&
+          ((tensor_proto->data_type() == ONNX_NAMESPACE::TensorProto_DataType_INT32) ||
+           (tensor_proto->data_type() == ONNX_NAMESPACE::TensorProto_DataType_INT64))) {
+        int64_t padding_idx = *reinterpret_cast<const int64_t*>(tensor_proto->raw_data().data());
+        if (padding_idx < 0) {
+          continue;
+        }
+        if (std::find(sparse_embedding_input_names_.begin(), sparse_embedding_input_names_.end(),
+                      node.InputDefs()[1]->Name()) == sparse_embedding_input_names_.end()) {
+          LOG_DEBUG_INFO(logger, "Skip node " + node.Name() + "(" + node.OpType() +
+                                     ") due to embedding input is not in the sparse embedding input list.");
+          continue;
+        }
+        embedding_node = &node;
+        input_ids_arg = embedding_node->MutableInputDefs()[1];
+        PushAllOutputNode(graph, to_visit, embedding_node, visited);
+        for (auto output_edges : embedding_node->MutableOutputDefs()) {
+          candidate_edges.insert(output_edges);
+        }
+        break;
+      }
+    }
+  }
+
+  if (!embedding_node) {
+    LOG_DEBUG_INFO(logger, "Exit PaddingElimination optimization for not finding any valid embedding node.");
+    return Status::OK();
+  }
+
+  while (!to_visit.empty()) {
+    Node* cur = to_visit.front();
+    to_visit.pop();
+    visited.insert(cur);
+    if (graph_utils::IsSupportedOptypeVersionAndDomain(*cur, "Add", {7, 13, 14})) {
```

Comment:
Looks we don't have unit test convering each operator, is the possible to add some?